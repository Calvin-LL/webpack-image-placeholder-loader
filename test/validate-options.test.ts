import webpack from "webpack";

import compile from "./helpers/compile";
import getCompiler from "./helpers/getCompiler";

describe("validate options", () => {
  const tests = {
    format: {
      success: ["base64", "hex", "rgb", "array"],
      failure: [true],
    },
    size: {
      success: [1, 100, "original"],
      failure: [0, -1],
    },
    color: {
      success: [
        "simple",
        "sqrt",
        "dominant",
        "#000",
        "#038281",
        "#03828110",
        { r: 2, g: 120, b: 24 },
        "red",
      ],
      failure: [true, "#0", "#MMMMMM"],
    },
    backgroundColor: {
      success: ["#000", "#038281", { r: 2, g: 120, b: 24 }, "red"],
      failure: [
        true,
        "#0",
        "#MMMMMM",
        "#03828110",
        { r: 2, g: 120, b: 24, a: 0.5 },
      ],
    },
    esModule: {
      success: [true, false],
      failure: ["true"],
    },
  };

  function createTestCase(
    key: string,
    value: any,
    type: "success" | "failure"
  ) {
    it(`should ${
      type === "success" ? "successfully validate" : "throw an error on"
    } the "${key}" option with ${JSON.stringify(value)} value`, async () => {
      const compiler = getCompiler({ [key]: value });

      let stats;

      try {
        stats = await compile(compiler);
      } finally {
        if (type === "success") {
          expect((stats as webpack.Stats).hasErrors()).toBe(false);
        } else if (type === "failure") {
          const {
            compilation: { errors },
          } = stats as any;

          expect(errors).toHaveLength(1);
          expect(() => {
            throw new Error(errors[0].error.message);
          }).toThrowErrorMatchingSnapshot();
        }
      }
    }, 60000);
  }

  for (const [key, values] of Object.entries(tests)) {
    for (const type of Object.keys(values) as ("success" | "failure")[]) {
      for (const value of values[type]) {
        createTestCase(key, value, type);
      }
    }
  }
});
