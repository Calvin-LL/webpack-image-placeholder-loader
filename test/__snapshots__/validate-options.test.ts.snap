// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate options should throw an error on the "backgroundColor" option with "#0" value 1`] = `
"(Emitted value instead of an instance of Error) Invalid options object. Image Placeholder Loader has been initialised using an options object that does not match the API schema.
      - options.backgroundColor \\"#0\\" is not a valid color."
`;

exports[`validate options should throw an error on the "backgroundColor" option with "#03828110" value 1`] = `
"(Emitted value instead of an instance of Error) Invalid options object. Image Placeholder Loader has been initialised using an options object that does not match the API schema.
      - options.backgroundColor \\"#03828110\\" is not a valid color."
`;

exports[`validate options should throw an error on the "backgroundColor" option with "#MMMMMM" value 1`] = `
"(Emitted value instead of an instance of Error) Invalid options object. Image Placeholder Loader has been initialised using an options object that does not match the API schema.
      - options.backgroundColor \\"#MMMMMM\\" is not a valid color."
`;

exports[`validate options should throw an error on the "backgroundColor" option with {"r":2,"g":120,"b":24,"a":0.5} value 1`] = `
"(Emitted value instead of an instance of Error) Invalid options object. Image Placeholder Loader has been initialised using an options object that does not match the API schema.
      - options.backgroundColor {\\"r\\":2,\\"g\\":120,\\"b\\":24,\\"a\\":0.5} is not a valid color."
`;

exports[`validate options should throw an error on the "backgroundColor" option with true value 1`] = `
"Invalid options object. Image Placeholder Loader has been initialized using an options object that does not match the API schema.
 - options.backgroundColor should be one of these:
   string | object { … }
   -> Color of background if image contains transparency.
   Details:
    * options.backgroundColor should be a string.
    * options.backgroundColor should be an object:
      object { … }"
`;

exports[`validate options should throw an error on the "color" option with "#0" value 1`] = `
"(Emitted value instead of an instance of Error) Invalid options object. Image Placeholder Loader has been initialised using an options object that does not match the API schema.
      - options.color \\"#0\\" is not a valid color."
`;

exports[`validate options should throw an error on the "color" option with "#MMMMMM" value 1`] = `
"(Emitted value instead of an instance of Error) Invalid options object. Image Placeholder Loader has been initialised using an options object that does not match the API schema.
      - options.color \\"#MMMMMM\\" is not a valid color."
`;

exports[`validate options should throw an error on the "color" option with true value 1`] = `
"Invalid options object. Image Placeholder Loader has been initialized using an options object that does not match the API schema.
 - options.color should be one of these:
   string | object { … }
   -> Color of output image.
   Details:
    * options.color should be a string.
    * options.color should be an object:
      object { … }"
`;

exports[`validate options should throw an error on the "esModule" option with "true" value 1`] = `
"Invalid options object. Image Placeholder Loader has been initialized using an options object that does not match the API schema.
 - options.esModule should be a boolean.
   -> By default, webpack-image-placeholder-loader generates JS modules that use the ES modules syntax."
`;

exports[`validate options should throw an error on the "format" option with true value 1`] = `
"Invalid options object. Image Placeholder Loader has been initialized using an options object that does not match the API schema.
 - options.format should be one of these:
   \\"base64\\" | \\"hex\\" | \\"rgb\\" | \\"array\\"
   -> Format of output"
`;

exports[`validate options should throw an error on the "size" option with -1 value 1`] = `
"Invalid options object. Image Placeholder Loader has been initialized using an options object that does not match the API schema.
 - options.size should be >= 1."
`;

exports[`validate options should throw an error on the "size" option with 0 value 1`] = `
"Invalid options object. Image Placeholder Loader has been initialized using an options object that does not match the API schema.
 - options.size should be >= 1."
`;
